pipeline/scripts/ci/upstream_cli.py

```diff
@@ -0,0 +1,160 @@
+        if exit_status:
+            raise Exception(f"Command failed to execute {cmd}")
+
+
+def store_in_upstream(branch_name, repo, image, version):
+    """Method to update it in upstream.yaml.
+    Args:
+        branch_name: Name of upstream branch
+        repo: upstream branch compose URL
+        image: upstream branch image
+        version: upstream branch ceph version.
+    """
+    url = "http://magna002.ceph.redhat.com/cephci-jenkins/latest-rhceph-container-info/upstream.yaml"
+    data = requests.get(url, timeout=30)
+    repo_details = yaml.safe_load(data.text)
+    existing_version = repo_details[branch_name]["ceph-version"]
+    # update only if current version is greater than existing version.
+    if version > existing_version:
+        update_build(branch_name, repo, image, version)
+
+
+def update_build(branch_name, repo, image, version):
+    # Mount reesi to update changes.
+    commands = [
+        "sudo mkdir -p /ceph",
+        "sudo mount -t nfs -o sec=sys,nfsvers=4.1 reesi004.ceph.redhat.com:/ /ceph",
+    ]
+    exec_sys_commands(commands)
+    file_name = "/ceph/cephci-jenkins/latest-rhceph-container-info/upstream.yaml"
+    try:
+        stream = open(file_name, "r")
+        data = yaml.load(stream)
+        data[branch_name]["composes"] = repo
+        data[branch_name]["image"] = image
+        data[branch_name]["ceph-version"] = version
+
+        with open(file_name, "w") as yaml_file:
+            yaml_file.write(yaml.dump(data, default_flow_style=False))
+        LOG.info(f"Updated build info in {file_name}")
+    except IOError as err:
+        LOG.error(f"IO error hit during opening the file. Error : {err}")
+
+
+usage = """
+This script helps to provide needed info for upstream specific.
+
+Usage:
+  upstream_cli.py build <branch_name>
+
+
+Example:
+    python upstream_cli.py build quincy
+
+Options:
+    -h --help                       Show this screen
+"""
+
+
+def fetch_upstream_build(branch: str):
+    """
+    Retuns the provided upstream branch rpm repo path and container image.
+
+    Args:
+        branch: str    Upstream branch name
+    """
+    branch = branch.lower()
+    branch_repo = branch + "/latest/centos/8/repo/"
+    url = "https://shaman.ceph.com/api/repos/ceph/" + branch_repo
+
+    # if any connection issue to URL it will wait for 30 seconds to establish connection.
+    response = requests.get(url, timeout=30)
+    if response.status_code != 200:
+        response.raise_for_status()
+    repo = response.url
+    LOG.info(f"Upstream repo for the given branch is {repo}")
+
+    # To get image for that rpm repo
+    id = re.search("[0-9a-f]{5,40}", repo).group()
+    image = "quay.ceph.io/ceph-ci/ceph:" + id
+    LOG.info(f"Upstream image for the given branch is {image}")
+
+    cmd = "podman -v"
+    exit_status = os.system(cmd)
+    if exit_status:
+        raise Exception("Please install podman on host")
+
+    cmd = f"sudo podman pull {image}"
```

Comment:
To confirm that image is valid from the registry, cross validating by pulling the image,
Bcz seen some of the images not able to pull for manifest unknown
`sangadi@sangadi ~]$ sudo podman pull quay.ceph.io/ceph-ci/ceph:ca74598065096e6fcbd8433c8779a2be0c889351
[sudo] password for sangadi: 
Trying to pull quay.ceph.io/ceph-ci/ceph:ca74598065096e6fcbd8433c8779a2be0c889351...
Error: initializing source docker://quay.ceph.io/ceph-ci/ceph:ca74598065096e6fcbd8433c8779a2be0c889351: reading manifest ca74598065096e6fcbd8433c8779a2be0c889351 in quay.ceph.io/ceph-ci/ceph: manifest unknown: manifest unknown`
So that instead moving further we can **Fail** here itself.

Tried pulling the image from a normal user it got failed as below,


> [sangadi@sangadi ~]$ podman pull quay.ceph.io/ceph-ci/ceph:f566756bcbf81a9e17ed272afd5e8555c4ded09b

ERRO[0000] cannot find UID/GID for user sangadi: No subuid ranges found for user "sangadi" in /etc/subuid - check rootless mode in man pages. 
WARN[0000] using rootless single mapping into the namespace. This might break some images. Check /etc/subuid and /etc/subgid for adding sub*ids 
Trying to pull quay.ceph.io/ceph-ci/ceph:f566756bcbf81a9e17ed272afd5e8555c4ded09b...
Getting image source signatures
Copying blob fee18ea417d6 done  
Copying blob 9f45cf1cd9f2 done  
Copying blob f0a2109a2528 done  
Copying blob 32033a8ea5a0 done  
Copying blob bfd1401568a8 done  
Error: writing blob: adding layer with blob "sha256:f0a2109a2528cd1cf5eaa3f5867a6771e4a7c28ae7179d7cc67cbcc2b29f21fc": Error processing tar file(exit status 1): potentially insufficient UIDs or GIDs available in user namespace (requested 0:5 for /usr/bin/write): Check /etc/subuid and /etc/subgid: lchown /usr/bin/write: invalid argument

using **sudo** able to pull the image successfully. hence used as sudo.
@psathyan Please let me know are you Ok with this?

Reply:
Thanks @sunilangadi2 for clarifying.

It would be great if you could add a documentation/comment about this at the top level. This would give us the reason why we need to run this script with root level privileges.